### PR TITLE
Build Docker images from exported contexts and test them

### LINKED ISSUE #69928: Add testing for generated Docker build contexts
We have the following tasks for generating Docker build contexts:

   * `distribution:docker:docker-build-context:buildDockerBuildContext`
   * `distribution:docker:ironbank-docker-build-context:buildIronBankDockerBuildContext`
   * `distribution:docker:ubi-docker-build-context:buildUbiDockerBuildContext`

However, we perform no testing on these to verify that the generated contexts are good, which has resulted in issue such as #69924.

We need to add testing for the default and UBI contexts, where we at a minimum build the images, and ideally run tests against them.

If possible, we should do the same for the Iron Bank context. The added difficulty here is that the Iron Bank context is intended to be build with a custom process. It ought to be straightforward to simulate this, however.

### LINKED ISSUE #69930: Simplify Docker build context tasks
We have the following tasks for generating Docker build contexts:

   * `distribution:docker:docker-build-context:buildDockerBuildContext`
   * `distribution:docker:ironbank-docker-build-context:buildIronBankDockerBuildContext`
   * `distribution:docker:ubi-docker-build-context:buildUbiDockerBuildContext`

Each of these is defined in its own sub-project, but this arrangement makes it unnecessarily fiddly to share logic between the main `docker` project and these sub-projects.

We should get rid of the sub-projects, perhaps defining these tasks in the `docker` project directly. This will have an impact on the release-manager, so we'll have to make changes there and coordinate with Infra.

### PR DESCRIPTION
**Pull request by pugnascotia**
Closes #69930. Closes #69928.

The ES build currently has 2 types of Docker output - Docker images, and Docker build contexts. At the moment, only the images are tested, meaning that bugs in the build contexts can go unnoticed.

This PR changes how we create Docker images so that we first create the build contexts, and then build the images using them. This does require some sleight-of-hand - the build contexts expect to download an Elasticsearch archive directorly from the `Dockerfile`, which will only ever work for non-snapshot version builds. In order to get around this, the `Dockerfile` is modified to `COPY` in a local archive file. Any other dependency files must exist in the build context archive.

This PR also builds and tests the Iron Bank context. We do not currently build a Docker image for this at all, and to build an image requires us to set some build arguments to useful values. We also need to provide all artifacts to the build, as the `Dockerfile` cannot download anything. As a result, the `:distribution:docker` project now defines a GitHub repository so that Gradle will download a `tini` binary.

Note that there will need to be corresponding changes to `release-manager`.

### CODE REVIEW

File: distribution/docker/build.gradle [hunk 1]
```diff
@@ -273,8 +304,91 @@ void addBuildDockerImage(Architecture architecture, DockerBase base) {
...
+        from configurations.transformLog4jJar
+      }
+    }
+
+  tasks.named("assemble").configure {
+    dependsOn(buildContextTask)
+  }
+
+  final TaskProvider<Copy> extractTask =
+    tasks.register(taskName('extract', architecture, base, 'DockerContext'), Copy) {
```
[the remark this answers is not in the document]
Comment (mark-vieira) [reply distribution/docker/build.gradle:336]: Here's another thing we should do. We should structure this such that building the context is a) copying all the stuff to a directory and b) then creating the archive.

The reason being is that currently what we do is build an archive, then unpack it, then build the docker image. That packing/unpacking step is not free. We did a bunch of work to do something similar for distributions for testing. So I would make the "build*DockerContext" task the `Copy` bit, and then create a`build*DockerContextTar` which we only care about for release manager. All the stuff internally that depends on the context will use teh `Copy` task, skipping the need to build the tarball altogether.

File: buildSrc/src/main/java/org/elasticsearch/gradle/docker/DockerBuildTask.java [hunk 2]
```diff
@@ -42,6 +43,7 @@
     private boolean pull = true;
     private boolean noCache = true;
     private String[] baseImages;
+    private MapProperty<String, String> buildArgs = getProject().getObjects().mapProperty(String.class, String.class);
```
[the remark this answers is not in the document]
Comment (pugnascotia) [reply buildSrc/src/main/java/org/elasticsearch/gradle/docker/DockerBuildTask.java]: I was following the existing `Project` references, but I've now used `ObjectFactory` throughout.

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/DistroTestPlugin.java [hunk 3]
```diff
@@ -223,7 +226,10 @@ public void apply(Project project) {
     private static Map<ElasticsearchDistribution.Type, TaskProvider<?>> lifecycleTasks(Project project, String taskPrefix) {
         Map<ElasticsearchDistribution.Type, TaskProvider<?>> lifecyleTasks = new HashMap<>();
         lifecyleTasks.put(Type.DOCKER, project.getTasks().register(taskPrefix + ".docker"));
-        lifecyleTasks.put(Type.DOCKER_UBI, project.getTasks().register(taskPrefix + ".ubi"));
+        lifecyleTasks.put(Type.DOCKER_CONTEXT, project.getTasks().register(taskPrefix + ".docker-context"));
```
[the remark this answers is not in the document]
Comment (pugnascotia) [reply buildSrc/src/main/java/org/elasticsearch/gradle/test/DistroTestPlugin.java]: I've pared this back so we're only adding one more for Iron Bank.

File: distribution/docker/build.gradle [hunk 4]
```diff
@@ -100,6 +119,18 @@ class SquashNewlinesFilter extends FilterReader {
   }
 }
 
+class ReplaceRemoteUrlWithLocalFilter extends FilterReader {
```
[the remark this answers is not in the document]
Comment (pugnascotia) [reply distribution/docker/build.gradle]: Part of the point of this work is specifically to take the context that we would sent to Docker, and use that to build our images. Templating it in the Dockerfile was one reason we had a proliferation of tasks and parameters.

File: distribution/docker/docker-aarch64-from-export-export/build.gradle [hunk 5]
```diff
@@ -0,0 +1,2 @@
+// This file is intentionally blank. All configuration of the
```
[the remark this answers is not in the document]
Comment (pugnascotia) [reply distribution/docker/docker-aarch64-from-export-export/build.gradle]: The `*-export` directories have been fixed up and pared back.

File: distribution/docker/build.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (pugnascotia) [reply distribution/docker/build.gradle]: I haven't (yet) done what you've suggested, but what we now have is:

Generate build context :arrow_right: assemble local build context :arrow_right: build image

The first archive is really small, so I realise it's not free to unpack, but it's a vanishingly small amount of time compared to building the Docker image.

In doing this, I've been able to simplify the logic.

File: distribution/docker/build.gradle (hunk 4 above)
[the remark this answers is not in the document]
Comment (mark-vieira) [reply distribution/docker/build.gradle]: Isn't the task that builds the context already doing _some_ replacements in that template though? We wouldn't have to add any additional tasks, would we?

File: distribution/docker/build.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (mark-vieira) [reply distribution/docker/build.gradle]: Sounds good. Yeah, I think I was under the impression the context would include the ES distribution itself as well, but it doesn't, which makes sense.

File: distribution/docker/build.gradle (hunk 4 above)
[the remark this answers is not in the document]
Comment (pugnascotia) [reply distribution/docker/build.gradle]: The task that builds the context does indeed perform a number of replacements, but these are all related to the type of image that is being built. It's not possible to generate a single Dockerfile that can switch between a local ES archive and a remote one, so we either modify the context's Dockerfile, or we need to have a way of parameterising how the context is built, ideally without requiring a separate task.

**Comment by pugnascotia** [comment]
@elasticmachine run elasticsearch-ci/packaging-unix

**Review by mark-vieira — APPROVED** [review]
LGTM, but keep in mind the CI checks here are not yet executing these new tasks. CI runs `destructiveDistroTest.docker`, we'll need to add additional jobs for `docker-ubi` and `docker-ironbank` for the new lifecycle tasks. I presume you've run these locally at least to verify they work.

**Comment by pugnascotia** [comment]
I've run all the tests locally and flushed out some issues. I also added another test for the images' `HEALTHCHECK` value.

**Comment by mark-vieira** [comment]
Sounds good, let's make a note/issue to follow up with a PR to update the packaging test job definitions.

**Comment by pugnascotia** [comment]
Backported to `7.x` in 1955f7181db, which also showed up a bunch of broken stuff for Iron Bank because we weren't building and testing the image (doesn't affect the actual Iron Bank repo, BTW).

**Pull request merged**